engine/pike2/pike2.go

```diff
@@ -0,0 +1,207 @@
+
+	// Verify that the resulting point is on the curve
+	if !curve.IsOnCurve(dedicatedPubKeyX, dedicatedPubKeyY) {
+		return nil, fmt.Errorf("resulting public key is not on curve")
+	}
+
+	// Create the dedicated public key
+	dPK := &bec.PublicKey{
+		Curve: curve,
+		X:     dedicatedPubKeyX,
+		Y:     dedicatedPubKeyY,
+	}
+	return dPK, nil
+}
+
+// convertToCompressedPublicKey converts the dedicated public key to its compressed format
+// and then hashes it using SHA-256 followed by RIPEMD-160 (Hash160).
+// Returns the compressed and hashed public key as a byte slice.
+func convertToCompressedPublicKey(dPK *bec.PublicKey) []byte {
+	// Serialize the public key to compressed format
+	dPKBytes := dPK.SerialiseCompressed()
+
+	// Apply Hash160 (SHA-256 followed by RIPEMD-160) to the compressed public key
+	dPKBytesCompressed := crypto.Hash160(dPKBytes)
+
+	// Return the compressed and hashed public key
+	return dPKBytesCompressed
+}
+
+// GenerateLockingScriptsFromTemplates converts Pike outputs templates to scripts
+func GenerateLockingScriptsFromTemplates(outputsTemplate []PikeOutputsTemplate, senderPubKey, receiverPubKey *bec.PublicKey, reference string) ([]string, error) {
+	lockingScripts := make([]string, len(outputsTemplate))
+	parser := interpreter.DefaultOpcodeParser{}
+	pubSharedSecret := CalculatePubSharedSecret(senderPubKey)
+
+	for idx, output := range outputsTemplate {
+		templateScript, err := bscript.NewFromHexString(output.Script)
+		if err != nil {
+			return nil, fmt.Errorf("error creating script from hex string - %w", err)
+		}
+		parsedScript, err := parser.Parse(templateScript)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing template script - %w", err)
+		}
+
+		hmacResult, err := calculateHMAC(pubSharedSecret.SerialiseCompressed(), reference, idx)
+		if err != nil {
+			return nil, err
+		}
+
+		dPK, err := calculateDedicatedPublicKey(hmacResult, receiverPubKey)
+		if err != nil {
+			return nil, err
+		}
+
+		dPKBytesCompressed := convertToCompressedPublicKey(dPK)
+
+		// Create a new script with the dedicated public key
+		newScript := new(bscript.Script)
+		if err := newScript.AppendPushData(dPKBytesCompressed); err != nil {
+			return nil, fmt.Errorf("error appending public key hash to script: %w", err)
+		}
+
+		// Parse the public key hash script
+		pkhParsed, err := parser.Parse(newScript)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing public key hash script: %w", err)
+		}
+
+		// Replace OP_PUBKEYHASH with the actual public key hash
+		evaluated := make([]interpreter.ParsedOpcode, 0, len(parsedScript))
+		for _, op := range parsedScript {
+			switch op.Value() {
+			case bscript.OpPUBKEYHASH:
+				evaluated = append(evaluated, pkhParsed...)
+			case bscript.OpPUBKEY:
+				evaluated = append(evaluated, pkhParsed...) // Currently just appending the same key
+			default:
+				evaluated = append(evaluated, op)
+			}
```

Comment:
It would be nice to have the whole logic of replacing templating OP codes in a separate function in package `script/template`.
The function could look like that:
```golang
func Evaluate(script []byte, pubKey *bec.PublicKey) []byte
```
or instead it could accept and return `string` (I don't know which version would be more sufficient to use in this code.

Also it would be good to have unit tests for it.